### PR TITLE
Implement wnd parsing code

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/client/gui/gamewindow.h [hunk 1]
```diff
@@ -109,25 +109,25 @@ enum GameWindowMessage
...
     GWM_MIDDLE_UP,
-    GWM_11,
+    GWM_RIGHT_CLICK,
     GWM_MIDDLE_DRAG,
 
     // Right mouse
     GWM_RIGHT_DOWN,
     GWM_RIGHT_UP,
-    GWM_15,
+    GWM_MIDDLE_CLICK,
```
Comment (xezon) [inline src/game/client/gui/gamewindow.h]: Not GWM_RIGHT_CLICK here?

File: src/game/client/gui/gamewindow.h [hunk 2]
```diff
@@ -109,25 +109,25 @@ enum GameWindowMessage
...
     GWM_LEFT_UP,
-    GWM_7,
+    GWM_LEFT_CLICK,
     GWM_LEFT_DRAG,
 
     // Middle mouse
     GWM_MIDDLE_DOWN,
     GWM_MIDDLE_UP,
-    GWM_11,
+    GWM_RIGHT_CLICK,
```
Comment (xezon) [inline src/game/client/gui/gamewindow.h]: Not GWM_MIDDLE_CLICK here?

File: src/game/client/gui/headertemplate.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+};
+
+class HeaderTemplateManager
+{
+public:
+    HeaderTemplate *Find_Header_Template(Utf8String name);
+    GameFont *Get_Font_From_Template(Utf8String name);
+
+private:
+    std::list<HeaderTemplate *> m_templates;
```
Comment (xezon) [inline src/game/client/gui/headertemplate.h:36]: No code here to populate `m_templates` ?

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 4]
```diff
@@ -12,4 +12,1685 @@
...
+    inst_data->m_imageOffset.y = atoi(strtok(nullptr, " \t\n\r"));
+    return true;
+}
+
+bool Parse_Font(char *token, class WinInstanceData *inst_data, char *buffer, void *data)
+{
+    strtok(buffer, " ,\n\r\t");
+    char *str;
+
+    for (str = buffer; *str != '\"'; str++) {
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: Escaping " in ' ' is not necessary. Several times in this file.

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 5]
```diff
@@ -12,4 +12,1685 @@
...
+    if (g_theNameKeyGenerator != nullptr) {
+        inst_data->m_id = g_theNameKeyGenerator->Name_To_Key(inst_data->m_decoratedNameString.Str());
+    }
+
+    return true;
+}
+
+bool Parse_Bit_Flag(const char *flag_string, unsigned int *bits, const char **flag_list)
+{
+    unsigned char c = 0;
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: bit

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 6]
```diff
@@ -12,4 +12,1685 @@
...
+    return 0;
+}
+
+void Parse_Bit_String(const char *in_buffer, unsigned int *bits, const char **flag_list)
+{
+    char buffer[256];
+
+    strcpy(buffer, in_buffer);
+
+    if (strncmp(buffer, "NULL", 4)) {
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: `!= 0`

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 7]
```diff
@@ -12,4 +12,1685 @@
...
+    nullptr };
+
+bool Parse_Style(char *token, class WinInstanceData *inst_data, char *buffer, void *data)
+{
+    inst_data->m_style = 0;
+    Parse_Bit_String(buffer, &inst_data->m_style, s_windowStyleNames);
+    return true;
+}
+
+bool Parse_System_Callback(char *token, class WinInstanceData *inst_data, char *buffer, void *data)
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp:247]: `const char *token`? Many times in this file.

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 8]
```diff
@@ -12,4 +12,1685 @@
...
+                color = &inst_data->m_disabledText;
+            } else {
+                if (i != 2) {
+                    captainslog_debug("Undefined state for text color");
+                    return false;
+                }
+
+                color = &inst_data->m_hiliteText;
+            }
+        } else {
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: To me this looks like compiler shuffled conditions. Who would write it like this?

Normal human would write it as

```c++
if (i == 0)
{
  color = &inst_data->m_enabledText;
}
else if (i == 1)
{
  color = &inst_data->m_disabledText;
}
else if (i == 2)
{
  color = &inst_data->m_hiliteText;
}
else
{
  captainslog_debug("Undefined state for text color");
  return false;
}
```

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 9]
```diff
@@ -12,4 +12,1685 @@
...
+        } else if (strcmp(token, "COMBOBOXEDITBOXHILITEDRAWDATA") == 0) {
+            draw_data = &s_hiliteEditBoxDrawData[i];
+        } else if (strcmp(token, "COMBOBOXLISTBOXENABLEDDRAWDATA") == 0) {
+            draw_data = &s_enabledListBoxDrawData[i];
+        } else if (strcmp(token, "COMBOBOXLISTBOXDISABLEDDRAWDATA") == 0) {
+            draw_data = &s_disabledListBoxDrawData[i];
+        } else if (strcmp(token, "COMBOBOXLISTBOXHILITEDRAWDATA") == 0) {
+            draw_data = &s_hiliteListBoxDrawData[i];
+        } else {
+            captainslog_debug("ParseDrawData, undefined token '%s'\n", token);
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: Can remove \n

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 10]
```diff
@@ -12,4 +12,1685 @@
...
+                buffer[i] = '\0';
+                return;
+            }
+
+            i++;
+        }
+    }
+
+    captainslog_debug("Read_Until_Semicolon: ERROR - Read buffer overflow - input truncated.");
+    buffer[max_buf_len - 1] = 0;
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: '\0'

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 11]
```diff
@@ -12,4 +12,1685 @@
...
+    info->m_shutdownNameString = strtok(buffer, " \n\r\t");
+    info->m_shutdownFunc = g_theFunctionLexicon->Win_Layout_Init_Func(
+        g_theNameKeyGenerator->Name_To_Key(info->m_shutdownNameString.Str()), FunctionLexicon::TABLE_WIN_LAYOUT_SHUTDOWN);
+    return true;
+}
+
+void Read_Until_Semicolon(File *file, char *buffer, int max_buf_len)
+{
+    int i = 0;
+    bool not_space = true;
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: is_before_word

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 12]
```diff
@@ -12,4 +12,1685 @@
...
+        for (;;) {
+            in_file->Scan_String(str);
+
+            if (str.Compare("ENDLAYOUTBLOCK") == 0) {
+                break;
+            }
+
+            for (LayoutScriptParse *parse = s_layoutScriptTable; parse != nullptr && parse->name != nullptr; parse++) {
+                if (str.Compare(parse->name) == 0) {
+                    Read_Until_Semicolon(in_file, buffer, 2048);
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp:762]: Not 2096? That is the buffer size calling into this function.

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 13]
```diff
@@ -12,4 +12,1685 @@
...
+        return nullptr;
+    } else {
+        return *(s_stackPtr - 1);
+    }
+}
+
+bool Parse_Screen_Rect(char *token, char *buffer, int *x, int *y, int *width, int *height)
+{
+    GameWindow *window = Peek_Window();
+    int i1;
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: upper

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 14]
```diff
@@ -12,4 +12,1685 @@
...
+    int i2;
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: left

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 15]
```diff
@@ -12,4 +12,1685 @@
...
+    int i3;
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: bottom

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 16]
```diff
@@ -12,4 +12,1685 @@
...
+    int i4;
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: right

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 17]
```diff
@@ -12,4 +12,1685 @@
...
+    int i5;
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: creation_res_x

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 18]
```diff
@@ -12,4 +12,1685 @@
...
+    int i6;
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: creation_res_y

as per

```
    SCREENRECT = UPPERLEFT: 8 595,
                 BOTTOMRIGHT: 13 600,
                 CREATIONRESOLUTION: 800 600;
```

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 19]
```diff
@@ -12,4 +12,1685 @@
...
+        *x = i1;
+        *y = i2;
+    }
+
+    *width = i3 - i1;
+    *height = i4 - i2;
+    return true;
+}
+
+bool Parse_Data(void **data, char *type, char *buffer)
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp:899]: `const char *type` ?

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 20]
```diff
@@ -12,4 +12,1685 @@
...
+{
+    if (!strcmp(type, "VERTSLIDER") || !strcmp(type, "HORZSLIDER")) {
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: == 0, several times in this function

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 21]
```diff
@@ -12,4 +12,1685 @@
...
+        *data = &sData;
+    } else if (!strcmp(type, "SCROLLLISTBOX")) {
+        static _ListboxData lData;
+        memset(&lData, 0, sizeof(_ListboxData));
+        lData.m_listLength = atoi(strtok(buffer, " \t\n\r"));
+        lData.m_autoScroll = atoi(strtok(nullptr, " \t\n\r")) != 0;
+        lData.m_autoPurge = atoi(strtok(nullptr, " \t\n\r")) != 0;
+        lData.m_scrollBar = atoi(strtok(nullptr, " \t\n\r")) != 0;
+        lData.m_multiSelect = atoi(strtok(nullptr, " \t\n\r")) != 0;
+        lData.m_forceSelect = atoi(strtok(nullptr, " \t\n\r")) != 0;
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp:915]: columns missing?

```
      LISTBOXDATA = LENGTH: 100,
                    AUTOSCROLL: 1,
                    SCROLLIFATEND: 0,
                    AUTOPURGE: 0,
                    SCROLLBAR: 1,
                    MULTISELECT: 0,
                    COLUMNS: 2,
                    COLUMNSWIDTH: 50,
                    COLUMNSWIDTH: 50,
                    FORCESELECT: 0;
```

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 22]
```diff
@@ -12,4 +12,1685 @@
...
+        *data = &lData;
+    } else if (!strcmp(type, "ENTRYFIELD")) {
+        static _EntryData eData;
+        memset(&eData, 0, sizeof(_EntryData));
+        eData.m_maxTextLen = atoi(strtok(buffer, " \t\n\r"));
+        strtok(nullptr, " \t\n\r");
+        const char *str = strtok(nullptr, " \t\n\r");
+
+        if (str) {
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: != nullptr

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 23]
```diff
@@ -12,4 +12,1685 @@
...
+            int i = atoi(str);
+            eData.m_secretText = i != 0;
+
+            if (i) {
+                eData.m_secretText = true;
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp:929]: This is a duplicate assignment. It is already assigned above with
```
eData.m_secretText = i != 0;
```
Error?

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 24]
```diff
@@ -12,4 +12,1685 @@
...
+            }
+        } else {
+            eData.m_secretText = false;
+        }
+
+        const char *str2 = strtok(nullptr, " \t\n\r");
+
+        if (str2) {
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: != nullptr

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 25]
```diff
@@ -12,4 +12,1685 @@
...
+        }
+
+        *data = &eData;
+    } else if (!strcmp(type, "STATICTEXT")) {
+        static _TextData tData;
+        int i = atoi(strtok(buffer, " \t\n\r"));
+        tData.m_centered = i != 0;
+
+        if (i) {
+            tData.m_centered = true;
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp:954]: This is a duplicate assignment. It is already assigned above with
```
tData.m_centered = i != 0;
```
Error?

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 26]
```diff
@@ -12,4 +12,1685 @@
...
+        }
+
+        strtok(nullptr, " \t\n\r");
+        *data = &tData;
+    } else if (!strcmp(type, "RADIOBUTTON")) {
+        static _RadioButtonData rData;
+        rData.m_group = atoi(strtok(buffer, " \t\n\r"));
+        *data = &rData;
+    } else {
+        *data = 0;
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: nullptr

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 27]
```diff
@@ -12,4 +12,1685 @@
...
+        } else {
+            eData.m_secretText = false;
+        }
+
+        const char *str2 = strtok(nullptr, " \t\n\r");
+
+        if (str2) {
+            eData.m_numericalOnly = atoi(str2) == 1;
+            eData.m_alphaNumericalOnly = atoi(str2) == 2;
+            eData.m_asciiOnly = atoi(str2) == 3;
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp:940]: This does not appear to match what Zero Hour wnd uses:

```
        TEXTENTRYDATA = MAXLEN: 64,
                        SECRETTEXT: 0,
                        NUMERICALONLY: 0,
                        ALPHANUMERICALONLY: 0,
                        ASCIIONLY: 0;
```
But perhaps it refers to something else. Not sure.

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 28]
```diff
@@ -12,4 +12,1685 @@
...
+        }
+
+        inst_data->m_style |= GWS_RADIO_BUTTON;
+        return g_theWindowManager->Go_Go_Gadget_Radio_Button(
+            parent, status, x, y, width, height, inst_data, rb_data, inst_data->m_font, false);
+    } else if (strcmp(type, "CHECKBOX") == 0) {
+        inst_data->m_style |= GWS_CHECK_BOX;
+        return g_theWindowManager->Go_Go_Gadget_Checkbox(
+            parent, status, x, y, width, height, inst_data, inst_data->m_font, false);
+    } else if (!strcmp(type, "TABCONTROL")) {
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: == 0, several times

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 29]
```diff
@@ -12,4 +12,1685 @@
...
+    } else if (!strcmp(type, "PROGRESSBAR")) {
+        inst_data->m_style |= GWS_PROGRESS_BAR;
+        return g_theWindowManager->Go_Go_Gadget_Progress_Bar(
+            parent, status, x, y, width, height, inst_data, inst_data->m_font, false);
+    }
+
+    return window;
+}
+
+void Set_Window_Text(GameWindow *window, Utf8String text_label)
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp:1215]: Utf8String label_name

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 30]
```diff
@@ -12,4 +12,1685 @@
...
+{
+    if (text_label.Is_Empty()) {
+        return;
+    } else {
+        Utf16String str;
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: label_text

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 31]
```diff
@@ -12,4 +12,1685 @@
...
+        Utf16String str2;
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: label_name2

File: src/game/client/gui/headertemplate.h (hunk 3 above)
Comment (xezon) [reply src/game/client/gui/headertemplate.h:36]: Nevermind, I saw it elsewhere externally.

File: src/game/client/gui/gamewindow.h (hunk 1 above)
Comment (jonwil) [reply src/game/client/gui/gamewindow.h]: fixed.

File: src/game/client/gui/gamewindow.h (hunk 2 above)
Comment (jonwil) [reply src/game/client/gui/gamewindow.h]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 4 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 6 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp:247]: No, original is char *.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 8 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 10 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 9 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 11 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 13 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 12 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp:762]: 2048 is correct.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 14 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 15 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 16 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 17 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 18 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 19 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp:899]: Original is not const.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 20 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 21 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp:915]: This is correct to original (and works in-game).

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 22 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 23 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp:929]: No this is what the original does (don't ask me why)

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 24 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 25 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp:954]: No this is what the original does (don't ask me why)

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 26 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 27 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp:940]: This is correct to original (and works in-game).

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 28 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 30 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp (hunk 31 above)
Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

File: src/game/client/gui/gamewindowmanagerscript.cpp [hunk 32]
```diff
@@ -12,4 +12,1685 @@
...
+                child_data->m_disabledDrawData[i] = s_disabledSliderDrawData[i];
+                child_data->m_hiliteDrawData[i] = s_hiliteSliderDrawData[i];
+            }
+
+            GameWindow *child = slider->Win_Get_Child();
+
+            if (child != nullptr) {
+                child_data = child->Win_Get_Instance_Data();
+
+                for (int i = 0; i < 9; i) {
```
Comment (xezon) [inline src/game/client/gui/gamewindowmanagerscript.cpp]: ++i

Comment (jonwil) [reply src/game/client/gui/gamewindowmanagerscript.cpp]: fixed.

**Review by xezon — APPROVED** [review]
UI works.

**Pull request merged**